lib/Canonicalize/SimplifySwitch/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,234 @@
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipeline/RegisterLLVMPass.h"
+#include "revng/Pipes/Ranks.h"
+#include "revng/Support/FunctionTags.h"
+#include "revng/Support/IRHelpers.h"
+#include "revng/ValueMaterializer/DataFlowGraph.h"
+#include "revng/ValueMaterializer/MemoryOracle.h"
+#include "revng/ValueMaterializer/ValueMaterializer.h"
+
+#include "revng-c/Pipes/Kinds.h"
+#include "revng-c/Support/FunctionTags.h"
+#include "revng-c/Support/IRHelpers.h"
+#include "revng-c/TypeNames/ModelTypeNames.h"
+
+#include "SimplifySwitchPass.h"
+
+using pipeline::serializedLocation;
+namespace ranks = revng::ranks;
+using BiDFNode = BidirectionalNode<DataFlowNode>;
+
+static Logger<> SwitchOptLogger("switch-opt");
+
+namespace {
+class DataMemoryOracle final : public MemoryOracle {
+private:
+  RawBinaryView &BinaryView;
+  model::Architecture::Values Arch;
+  bool IsLittleEndian = false;
+
+public:
+  DataMemoryOracle(RawBinaryView &BinaryView,
+                   model::Architecture::Values Arch) :
+    BinaryView(BinaryView), Arch(Arch) {}
+  ~DataMemoryOracle() final = default;
+
+  MaterializedValue load(uint64_t LoadAddress, unsigned LoadSize) final {
+    auto Address = MetaAddress::fromGeneric(toLLVMArchitecture(Arch),
+                                            LoadAddress);
+    auto MaybeValue = BinaryView.readInteger(Address, LoadSize, IsLittleEndian);
+    auto NewAPInt = [LoadSize](uint64_t V) {
+      return llvm::APInt(LoadSize * 8, V);
+    };
+    if (MaybeValue)
+      return MaterializedValue::fromConstant(NewAPInt(*MaybeValue));
+    else
+      return MaterializedValue::invalid();
+  }
+};
+} // namespace
+
+static llvm::Constant *toLLVMConstant(llvm::LLVMContext &Context,
+                                      const llvm::APInt &Value) {
+  using namespace llvm;
+  llvm::Type *DestinationType = llvm::IntegerType::get(Context,
+                                                       Value.getBitWidth());
+  if (DestinationType->isPointerTy()) {
+    revng_abort();
+  } else if (auto *IT = dyn_cast<IntegerType>(DestinationType)) {
+    revng_assert(IT->getBitWidth() == Value.getBitWidth());
+    return ConstantInt::get(DestinationType, Value);
+  } else {
+    revng_abort();
+  }
+}
+
+static llvm::BasicBlock *getBlockFor(llvm::SwitchInst *Switch,
+                                     llvm::Constant *C) {
+  uint64_t TheConstant = cast<llvm::ConstantInt>(C)->getValue().getZExtValue();
+  for (const auto &I : Switch->cases()) {
+    const llvm::ConstantInt *CaseVal = I.getCaseValue();
+    uint64_t Case = CaseVal->getValue().getZExtValue();
+    if (Case == TheConstant)
+      return I.getCaseSuccessor();
+  }
+
+  return nullptr;
+}
+
+static bool runSimplifySwitch(llvm::Function &F,
+                              llvm::LazyValueInfo &LVI,
+                              llvm::DominatorTree &DT,
+                              DataMemoryOracle &MO) {
+  if (F.isDeclaration() || !F.hasExactDefinition())
+    return false;
```

Comment:
1. What does `hasExactDefinition` do?
2. Unnecessary.

Reply:
We usually skip those when writing LLVM Passes since some functions can have `derefined` definitions, but yeah, it does not apply to this use case too much.